kafka/consumer.go

```diff
@@ -0,0 +1,401 @@
+			}
+
+			if err == nil {
+				lag := int(newest - offset)
+				c.partitionLag[partition].Set(lag)
+				if c.LagMonitor != nil {
+					c.LagMonitor.StoreLag(partition, lag)
+				}
+			}
+		}
+	}
+
+	ticker := time.NewTicker(c.conf.OffsetCommitInterval)
+	for {
+		select {
+		case ts := <-ticker.C:
+			storeOffsets(ts)
+		case <-c.stopChan:
+			storeOffsets(time.Now())
+			return
+		}
+	}
+}
+
+func (c *Consumer) startConsumer() error {
+	var offset confluent.Offset
+	var err error
+	var topicPartitions confluent.TopicPartitions
+	c.currentOffsets = make(map[int32]*int64, len(c.Partitions))
+
+	for i, topic := range c.conf.Topics {
+		for _, partition := range c.Partitions {
+			var currentOffset int64
+			switch c.conf.StartAtOffset {
+			case "oldest":
+				currentOffset, err = c.tryGetOffset(topic, partition, int64(confluent.OffsetBeginning), 3, time.Second)
+				if err != nil {
+					return err
+				}
+			case "newest":
+				currentOffset, err = c.tryGetOffset(topic, partition, int64(confluent.OffsetEnd), 3, time.Second)
+				if err != nil {
+					return err
+				}
+			default:
+				offsetDuration, err := time.ParseDuration(c.conf.StartAtOffset)
+				if err != nil {
+					return fmt.Errorf("invalid offest format %s: %s", c.conf.StartAtOffset, err)
+				}
+				currentOffset = time.Now().Add(-1*offsetDuration).UnixNano() / int64(time.Millisecond)
+				currentOffset, err = c.tryGetOffset(topic, partition, currentOffset, 3, time.Second)
+				if err != nil {
+					log.Warn("kafka-consumer: Failed to get specified offset %s, falling back to \"oldest\"", c.conf.StartAtOffset)
+					currentOffset, err = c.tryGetOffset(topic, partition, int64(confluent.OffsetBeginning), 3, time.Second)
+					if err != nil {
+						return err
+					}
+				}
+			}
+
+			offset, err = confluent.NewOffset(currentOffset)
+			if err != nil {
+				return err
+			}
+
+			topicPartitions = append(topicPartitions, confluent.TopicPartition{
+				Topic:     &topic,
+				Partition: partition,
+				Offset:    offset,
+			})
+
+			if i == 0 {
+				c.currentOffsets[partition] = &currentOffset
+			}
+		}
+	}
+
+	return c.consumer.Assign(topicPartitions)
+}
+
+func (c *Consumer) tryGetOffset(topic string, partition int32, offsetI int64, attempts int, sleep time.Duration) (int64, error) {
+	offset, err := confluent.NewOffset(offsetI)
+	if err != nil {
+		return 0, err
+	}
+
+	var beginning, end int64
+
+	attempt := 1
+	for {
+		if offset == confluent.OffsetBeginning || offset == confluent.OffsetEnd {
+			beginning, end, err = c.consumer.QueryWatermarkOffsets(topic, partition, c.conf.MetadataTimeout)
+			if err == nil {
+				if offset == confluent.OffsetBeginning {
+					return beginning, nil
+				} else {
+					return end, nil
+				}
+			}
+		} else {
+			times := []confluent.TopicPartition{{Topic: &topic, Partition: partition, Offset: offset}}
+			times, err = c.consumer.OffsetsForTimes(times, c.conf.MetadataTimeout)
+			if err != nil || len(times) == 0 {
```

Comment:
The errors are not correctly being handled.   You are setting `offset` but it is never returned.
This whole section needs to be cleaned up to something like:
```
if err != nil {
  log.Error(3, "failed to get offset", err)
} else {
  if len(times) == 0 {
     log.Info("falling back to oldest due to %s",err)
     return confluent.OffsetBeginning, nil
  } 
  return int64(times[0].Offset), nil
}
```

Reply:
:+1: slightly modified, but mostly copy-paste: https://github.com/grafana/metrictank/pull/879/commits/6fe611828e5d9241bf1c88b41bc0e1de6a92b79b